tests/unit/unstable/objrestspread.js

```diff
@@ -0,0 +1,109 @@
+"use strict";
+
+var TestRun = require("../../helpers/testhelper").setup.testRun;
+
+exports.enabling = function (test) {
+  TestRun(test, "Not enabled")
+    .addError(1, "'object spread property' is a non-standard language feature. Enable it using the 'objspreadrest' unstable option.")
+    .test("void { ...x };", { esversion: 6 });
+
+  TestRun(test, "Not enabled")
+    .addError(1, "'object rest property' is a non-standard language feature. Enable it using the 'objspreadrest' unstable option.")
+    .test("({ ...x } = {});", { esversion: 6 });
+
+  test.done();
+};
+
+exports.spread = function (test) {
+  var code = [
+    "var o;",
+    "o = { ...o };",
+    "o = { set x(_) {}, ...o, get x() {} };",
+    "o = { *gen() { yield; }, ...o, [o]() {} };",
+    "o = { ...o, };"
+  ];
+
+  TestRun(test, "identifier")
+    .test(code, { esversion: 6, unstable: { objspreadrest: true } });
+
+  code = [
+    "var o;",
+    "o = { ...this };",
+    "o = { ...[] };",
+    "o = { ...'string' };",
+    "o = { ...o = {} };",
+    "o = { ...() => {} };",
+    "o = { ...o => {} };"
+  ];
+
+  TestRun(test, "expression")
+    .test(code, { esversion: 6, unstable: { objspreadrest: true } });
+
+  test.done();
+};
+
+exports.rest = function (test) {
+  var code = [
+    "({ ...x } = {});",
+    "({ a, ...x } = {});",
+    "({ a = 0, ...x } = {});",
+    "({ a: A, ...x } = {});",
+    "({ a: A = 0, ...x } = {});"
+  ];
+
+  TestRun(test, "identifier, final")
+    .test(code, { esversion: 6, unstable: { objspreadrest: true } });
+
+  code = [
+    "({ ...x, } = {});",
+    "({ a, ...x, b } = {});",
+    "({ a = 0, ...x, b = 1 } = {});",
+    "({ a: A, ...x, b: B } = {});",
+    "({ a: A = 0, ...x, b: B = 0 } = {});"
+  ];
+
+  TestRun(test, "identifier, not final")
+    .addError(1, "Invalid element after rest element.")
+    .addError(2, "Invalid element after rest element.")
+    .addError(3, "Invalid element after rest element.")
+    .addError(4, "Invalid element after rest element.")
+    .addError(5, "Invalid element after rest element.")
+    .test(code, { esversion: 6, unstable: { objspreadrest: true } });
+
+  code = [
+    "({ ...[a, b, c] } = {});",
+    "({ a, ...[b, c, d] } = {});",
+    "({ a = 0, ...[b, c, d] } = {});",
+    "({ a: A, ...[b, c, d] } = {});",
+    "({ a: A = 0, ...[b, c, d] } = {});"
```

Comment:
Per recent consensus, this is invalid: `{ ...[] } = {}`

- http://tc39.github.io/tc39-notes/2017-05_may-23.html#16ih-why-allow-bindingpattern-for-bindingrestparameter-for-object-rest-maybe-we-should-just-allow-identifiers
- http://tc39.github.io/tc39-notes/2017-05_may-23.html#conclusionresolution-16ih-why-allow-bindingpattern-for-bindingrestparameter-for-object-rest-maybe-we-should-just-allow-identifiers